peer/ping_manager.go

```diff
@@ -0,0 +1,229 @@
+package peer
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/lightningnetwork/lnd/lnwire"
+)
+
+type PingManagerConfig struct {
+
+	// NewPingPayload is a closure that returns the payload to be packaged
+	// in the Ping message.
+	NewPingPayload func() []byte
+
+	// NewPongSize is a closure that returns a random value between
+	// [0, lnwire.MaxPongBytes]. This random value helps to more effectively
+	// pair Pong messages with Ping.
+	NewPongSize func() uint16
+
+	// IntervalDuration is the Duration between attempted pings.
+	IntervalDuration time.Duration
+
+	// TimeoutDuration is the Duration we wait before declaring a ping
+	// attempt failed.
+	TimeoutDuration time.Duration
+
+	SendPing      func(ping *lnwire.Ping)
+	OnPongFailure func()
+}
+
+// pingManager is a structure that is designed to manage the internal state
+// of the ping pong lifecycle with the remote peer. We assume there is only one
+// ping outstanding at once.
+//
+// NOTE: This structure MUST be initialized with NewPingManager.
+type PingManager struct {
+	cfg *PingManagerConfig
+
+	// pingTime is a rough estimate of the RTT (round-trip-time) between us
+	// and the connected peer.
+	// To be used atomically.
+	// TODO(roasbeef): also use a WMA or EMA?
+	pingTime atomic.Pointer[time.Duration]
+
+	// pingLastSend is the time when we sent our last ping message.
+	// To be used atomically.
+	pingLastSend *time.Time
+
+	// outstandingPongSize is the current size of the requested pong
+	// payload.  This value can only validly range from [0,65531]. Any
+	// value < 0 is interpreted as if there is no outstanding ping message.
+	outstandingPongSize int32
+
+	// pingTicker is a pointer to a Ticker that fires on every ping
+	// interval.
+	pingTicker *time.Ticker
+
+	// pingTimeout is a Timer that will fire when we want to time out a
+	// ping
+	pingTimeout *time.Timer
+
+	// pongChan is the channel on which the pingManager will write Pong
+	// messages it is evaluating
+	pongChan chan *lnwire.Pong
+
+	started sync.Once
+	stopped sync.Once
+
+	quit chan struct{}
+	wg   sync.WaitGroup
+}
+
+// NewPingManager constructs a pingManager in a valid state. It must be started
+// before it does anything useful, though.
+func NewPingManager(cfg *PingManagerConfig) *PingManager {
+	m := PingManager{
+		cfg:                 cfg,
+		outstandingPongSize: -1,
+		pongChan:            make(chan *lnwire.Pong, 1),
+		quit:                make(chan struct{}),
+	}
+
+	return &m
+}
+
+// Start launches the primary goroutine that is owned by the pingManager.
+func (m *PingManager) Start() error {
+	var err error
+	m.started.Do(func() {
+		err = m.start()
+	})
+
+	return err
+}
+
+func (m *PingManager) start() error {
+	m.pingTicker = time.NewTicker(m.cfg.IntervalDuration)
+
+	m.pingTimeout = time.NewTimer(0)
+	defer m.pingTimeout.Stop()
+
+	if !m.pingTimeout.Stop() {
+		<-m.pingTimeout.C
+	}
```

Comment:
can you add a comment explaining what this is doing pls?

Reply:
could add a few more comments to this function in general

Reply:
Yeah I can. This is done to ensure that the timeout channel is drained and in the empty state. This is done so that we can allocate a single timer at the start of the running lifecycle of the PingManager and just subsequently reset it each time we need to start the clock on a new timeout.

It was taken directly from the [golang Timer docs](https://pkg.go.dev/time#Timer.Stop)